File: apps/csharp/src/Ch06/GameLoop.cs
namespace GrokkingConcurrency.CSharp.Ch06;

/// <summary>
/// タスクを表すレコード
/// </summary>
public record GameTask(string Name, Action Action)
{
    public void Run() => Action();
}

/// <summary>
/// プロセッサ空き状態を示すイベント
/// threading.Event の C# 版
/// </summary>
public class ProcessorFreeEvent
{
    private readonly object _lock = new();
    private bool _signaled;

    /// <summary>
    /// シグナルを待機
    /// </summary>
    public void WaitForSignal()
    {
        lock (_lock)
        {
            while (!_signaled)
            {
                Monitor.Wait(_lock);
            }
            _signaled = false; // シグナルを消費
        }
    }

    /// <summary>
    /// シグナルを送信
    /// </summary>
    public void Signal()
    {
        lock (_lock)
        {
            _signaled = true;
            Monitor.Pulse(_lock);
        }
    }
}

/// <summary>
/// ゲームループ - マルチタスキングの例
/// 入力処理、ゲーム計算、描画処理を順番に実行
/// </summary>
public class GameLoop(
    Action inputTask,
    Action computeTask,
    Action renderTask,
    Func<bool> continueCondition)
{
    /// <summary>
    /// ゲームループを実行
    /// </summary>
    public void Run()
    {
        while (continueCondition())
        {
            inputTask();
            computeTask();
            renderTask();
        }
    }
}

File: apps/fsharp/src/Ch06/GameLoop.fs
namespace GrokkingConcurrency.Ch06

open System.Threading

type GameTask = { Name: string; Action: unit -> unit }

type ProcessorFreeEvent = {
    Lock: obj
    mutable Signaled: bool
}

module GameLoop =
    /// Run a single task
    let runTask (task: GameTask) : unit =
        task.Action()

    /// Create a new processor free event
    let createProcessorFreeEvent () : ProcessorFreeEvent =
        { Lock = obj(); Signaled = false }

    /// Wait for the event to be signaled
    let waitForSignal (event: ProcessorFreeEvent) : unit =
        lock event.Lock (fun () ->
            while not event.Signaled do
                Monitor.Wait(event.Lock) |> ignore
        )

    /// Signal the event
    let signal (event: ProcessorFreeEvent) : unit =
        lock event.Lock (fun () ->
            event.Signaled <- true
            Monitor.PulseAll(event.Lock)
        )

    /// Reset the event
    let reset (event: ProcessorFreeEvent) : unit =
        lock event.Lock (fun () ->
            event.Signaled <- false
        )

    /// Run all tasks in a single frame
    let runOneFrame (tasks: GameTask list) : unit =
        tasks |> List.iter runTask

    /// Run game loop for specified number of frames
    let run (tasks: GameTask list) (frames: int) : unit =
        for frame in 1 .. frames do
            printfn $"Frame {frame}"
            runOneFrame tasks
